### DISCUSSION TITLE
Drag drop sort- no animation on drag over

### DISCUSSION THREAD

**Discussion by GradySain** [clauderic/dnd-kit#1068] [Q&A]
Hello,

I didn't want to open an issue on this because I'm probably just doing something wrong... This video has audio- turning it you'll hear my description of the issue...

https://user-images.githubusercontent.com/97621176/225139074-0e5201e4-e571-4446-a512-fd2c5f3d5975.mp4

Here's the source:

https://stackblitz.com/edit/react-zsldgk?file=src/App.js

I'm trying to create setup that allows a user to drag items from the left side bar, and drop it onto the drop zone. Items can be re-used, dragged any number of times to the drop zone. When dragging over the drop zone, new "temporary" items should automatically be created (if they are dragged over existing items), and automatically sorted until they are finally dropped.

Where I'm stuck- I can't get the sorting animation to work when items are being dragged over the drop zone. Once items are in the drop zone, they can be dragged, and the sorting animation works correctly.

There are many things "wrong" in the current state of the code. Before I go on to correct those issues, I need to understand why why the sorting animation isn't working. I commented out the use of DragOverlay on purpose, again to understand why the current state of the code isn't working...

If anyone has any hints or clues as to what I'm doing wrong here, I'd be grateful for input. I built this example using many other existing examples, trying to get the right combinations of features for my setup.

**Comment by lukewatts** [comment]
Having similar issues with this library. Seems like a lot has to be manually handled in the onDragOver to get things to work, which isn;t ideal. Even in their examples they have code that checks if the items is over the first item or `isBelowLastItem`. Seems to me like it should just handle this internally, at least from an animation point of view.

Anyways, it might be something like this code that you're missing...everything from `let newIndex`:
 https://github.com/clauderic/dnd-kit/blob/master/stories/2%20-%20Presets/Sortable/MultipleContainers.tsx#L336

But honestly, I think this library is way too complex for even simple use cases.

**Reply by GradySain** [reply]
@lukewatts Thanks for taking a look.  I've seen examples that handle the exact functionality I'm missing, but I can't seem to find the correct, non-hacky way to implement. `OnDragOver `examples I've seen don't really have any extra logic in them to "create" animation... I know I'm missing something fundamental here...

@clauderic any way you could take a look at my code and see if something jumps out at you?

**Comment by GradySain** [comment]
@clauderic I'm still stuck on this issue. Is there any way you could take a look and see if I'm doing something wrong here?

**Reply by pstrassmann** [reply]
The animations that govern sorting and animation come as return values from the `useSortable` hook. Specifically, `transform` governs where the items will be moved to (by css), and `transition` governs the animation of that css transform. The values of `transform` and `transition` should then be passed as CSS to the things that should be sortable. Dndkit manages the value of `transition` for all sortable items so that things only animate when appropriate.

However, for these to work correctly it's important that the `items` prop for `SortableContext` is always in sync with the order that things are getting rendered in. When you're using the draggable to resort items in the drop zone, are `items` being correctly updated? That's the first thing that comes to mind for me

**Comment by GradySain** [comment]
The fix I ended up using. In App.js (in the stackblitz in original post), replace these lines (111-117)

```
const temp = [...droppedItems];
      const lastAddedID = temp.length;
      setDroppedItems((items) => {
        const oldIndex = items.findIndex((item) => item.sortID === lastAddedID);
        const newIndex = items.findIndex((item) => item.sortID === over.id);
        return arrayMove(items, oldIndex, newIndex);
      });
```
with this:

```
// found it here: https://codesandbox.io/s/github/jdthorpe/dnd-kit-sortable-poc?file=/src/App.tsx:3398-3509
      if (oldIndex === newIndex) return;
      // also critical, seems to be against react rules, directly mutating state, but works...
      setDroppedItems(arrayMove(droppedItems, newIndex, oldIndex));
```
      
I don't use approaches like this anywhere else in my code. But for some reason this, and only this, seemed to work. Have no idea why.
